versions/development/SPEC.md

```diff
@@ -344,20 +345,79 @@ Tasks and workflow inputs may be passed in from an external source, or they may
 
 A string literal may contain any unicode characters between single or double-quotes, with the exception of a few special characters that must be escaped:
 
-| Escape Sequence | Meaning | \x Equivalent | Context |
-|----|---|------|--|
-|`\\`|`\`|`\x5C`||
-|`\n`|newline|`\x0A`||
-|`\t`|tab|`\x09`||
-|`\'`|single quote|`\x27`|within a single-quoted string|
-|`\"`|double quote|`\x22`|within a double-quoted string|
+| Escape Sequence | Meaning      | \x Equivalent | Context                       |
+| --------------- | ------------ | ------------- | ----------------------------- |
+| `\\`            | `\`          | `\x5C`        |                               |
+| `\n`            | newline      | `\x0A`        |                               |
+| `\t`            | tab          | `\x09`        |                               |
+| `\'`            | single quote | `\x27`        | within a single-quoted string |
+| `\"`            | double quote | `\x22`        | within a double-quoted string |
 
 Strings can also contain the following types of escape sequences:
 
 * An octal escape code starts with `\`, followed by 3 digits of value 0 through 7 inclusive.
 * A hexadecimal escape code starts with `\x`, followed by 2 hexadecimal digits `0-9a-fA-F`. 
 * A unicode code point starts with `\u` followed by 4 hexadecimal characters or `\U` followed by 8 hexadecimal characters `0-9a-fA-F`.
 
+##### Multi-line Strings
+
+Strings that begin and end with three consecutive single- or double-quotes may span multiple lines. The opening quotes of a multi-line string may be followed by whitespace (optional) and must have a newline before the first non-whitespace character; these leading whitespace/newline characters are removed. All subsequent non-empty lines are then used to determine the multi-line string's *common leading whitespace* - the minimum number of whitespace characters occuring before the first non-whitespace character in a line or the end of the line, whichever comes first. This common leading whitespace is stripped from the beginning of all the lines in the multi-line string.
```

Comment:
> must have a newline before the first non-whitespace character

This 'must' is a little surprising, where does it come from? NBD but sometimes it's useful to write one-line string literals that contain both single and double quotes, e.g. in python:

```
>>> print('''He continued in another tweet on Aug. 29: "When you see 'anonymous source,' stop reading the story, it is fiction!"''')
He continued in another tweet on Aug. 29: "When you see 'anonymous source,' stop reading the story, it is fiction!"
```

Reply:
@mlin this comes from the Java specification which is the inspiration for the multi line strings

Reply:
That makes sense. In python the multi-line strings must be dedented using the `textwrap.dedent` function if there is leading whitespace. I do not want that sort of helper functions to be necessary in WDL. So java-style multiline strings that auto-strip the indentation make the most sense to me. Even if that means there are extra requirements to writing the multiline string.

Reply:
We could provide the behavior @mlin is looking for at the expense of making the specification more complex. We'd have to decide what behavior we want in the case of a multi-line string without a newline before the first non-whitespace:

1. Don't consider the first line when determining common leading whitespace
2. Determine leading whitespace the same way for the first line as subsequent lines

Consider the following examples:

```wdl
String s1 = """This is
    a multi-line string"""
String s2 = """  This is
    a multi-line string"""
String s3 = """    This is
    a multi-line string"""
```

If we go with option 1, then the common leading whitespace is 4 for for all three strings. Also, in `s2` and `s3`, do we remove the leading whitespace from the first line or not?

If we go with option 2, then the common leading whitespace is 0 for `s1`, 2 for `s2`, and 4 for `s3`.

I don't think either of these options is particularly intuitive, but option 2 would (IMHO) be less surprising behavior than option 1.

Reply:
I'm not going to veto anything here, just noting I was surprised